src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -258,25 +248,59 @@ class ResourceHashtableBase : public STORAGE {
   template<class ITER>
   void unlink(ITER* iter) {
     const unsigned sz = table_size();
-    for (unsigned index = 0; index < sz; index++) {
+    int cnt = _number_of_entries;
+
+    for (unsigned index = 0; cnt > 0 && index < sz; index++) {
       Node** ptr = bucket_at(index);
+
       while (*ptr != nullptr) {
         Node* node = *ptr;
         // do_entry must clean up the key and value in Node.
         bool clean = iter->do_entry(node->_key, node->_value);
+
         if (clean) {
           *ptr = node->_next;
           if (ALLOC_TYPE == AnyObj::C_HEAP) {
             delete node;
           }
-          _number_of_entries --;
+          _number_of_entries--;
         } else {
           ptr = &(node->_next);
         }
+        if (--cnt <= 0) return;
       }
     }
   }
 
+  // unlink_all() is a specialized version of unlink() when we decide to remove all elements.
+  // It can not replace unlink(ITER* iter) if user-provided iter releases key/value
+  void unlink_all() {
+    Node** bucket = table();
+    const unsigned sz = table_size();
+
+    while (_number_of_entries > 0 && bucket < bucket_at(sz)) {
+      Node* node = *bucket;
```

Comment:
Instead of having two loops for unlinking, I would suggest using the same pattern as iterate/iterate_all so you can share the same loop for the different APIs. This will also allow you to use C++ Lambda for unlinking.

Reply:
hi, Lam
Are you suggesting to use lambda+iterate() to implement unlink_all()? If we use `bucket` and `node` captured by the lambda, I think it's possible. However,  I think the resultant code is much less reasonable. Is it worth it? 

thanks,
--lx

Reply:
I mean we should implement a basic version of unlink() like this

```
template<typename Function>
void unlink(Function function) const { ...  }
```

Then `unlink_all` and  `template<class ITER> void unlink(ITER* iter)` can call into the above template.

Reply:
hi, @iklam @coleenp, 
I added `unlink(Function f)` per your request. Could you take a look?